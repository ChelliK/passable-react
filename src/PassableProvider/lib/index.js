import inputAttributesByType from './inputAttributesByType'

export {inputAttributesByType};