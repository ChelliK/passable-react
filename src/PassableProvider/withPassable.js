import React from 'react';
import PropTypes from 'prop-types';

export default function withPassable(PassableEnabledForm) {
    const Provide = (props, context) => (
        <PassableEnabledForm {...props} {...context}/>
    );

    Provide.contextTypes = {
        onChange: PropTypes.func,
        onBlur: PropTypes.func,
        fields: PropTypes.object,
        warnings: PropTypes.object,
        errors: PropTypes.object
    };

    return Provide;
}