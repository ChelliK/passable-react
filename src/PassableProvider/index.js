import React, { Component } from 'react';
import passable from 'passable';
import {inputAttributesByType, buildFieldsObject, mergeFieldIntoState} from './lib'

class PassableProvider extends Component {
    constructor(props) {
        super(props);

        this.state = {
            fields: this.initFields(props.initialFormState),
            errors: {},
            warnings: {}
        };

        this.onBlur = this.onBlur.bind(this);
        this.onChange = this.onChange.bind(this);
        this.custom = this.props.custom || {};
        this.passes = this.props.passes;
    }

    initFields(initialFormState = {}) {
        return buildFieldsObject(initialFormState);
    }

    processResults(passableObject) {
        this.setState((prevState) => {

        });
    }

    validate(specific = []) {
        const formData = this.getFields();
        const result = passable(this.props.name,
            specific,
            this.passes(formData),
            this.custom);

        this.processResults(result);
    }

    getFields() {
        return Object.assign({}, this.state.fields);
    }

    getFieldAttributes(name, element) {
        if (!element) {
            return {};
        }

        const formData = this.getFields();

        return Object.assign({}, formData[name], inputAttributesByType(element));
    }

    onChange(e) {
        const target = e.target;
        const { name } = target;
        const fieldAttributes = this.getFieldAttributes(name, target);

        this.setInField(name, {
            dirty: true,
            ...fieldAttributes
        });

        this.validate(name);

    }

    onBlur(e) {
        const target = e.target;
        const { name } = target;
        const fieldAttributes = this.getFieldAttributes(name, target);

        this.setInField(name, {
            touched: true,
            ...fieldAttributes
        });
    }

    setInField(name, entries = {}) {
        this.setState((prevState) => mergeFieldIntoState(prevState, name, entries));
    }

    render() {
        const { children } = this.props;

        return (
            children({
                onChange: this.onChange,
                onBlur: this.onBlur,
                fields: this.state.fields,
                errors: this.state.errors,
                warnings: this.state.warnings
            })
        );
    }
}

export default PassableProvider;