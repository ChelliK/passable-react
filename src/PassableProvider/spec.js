import React from 'react';
import {shallow} from 'enzyme';
import PassableProvider from './index';