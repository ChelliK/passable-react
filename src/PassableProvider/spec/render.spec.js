import React from 'react';
import { shallow } from 'enzyme';
import PassableProvider from '../index';

describe('Test PassableProvider render function', () => {

    const initialFormState = {
        field_1: 'hello',
        field_2: {
            checked: true
        }
    }
    const formRender = shallow(
        <PassableProvider initialFormState={initialFormState}>
            {(fields) => (
                <form action="">
                    <label className="username">Username
                        <input type="text" defaultValue={initialFormState.field_1} name="field_1"/>
                    </label>
                    <label className="is-alive">Is Alive
                        <input type="checkbox" defaultChecked={initialFormState.field_2.checked} name="field_2"/>
                    </label>
                </form>
            )}
        </PassableProvider>
    );

    describe('PassableProvider children rendering', () => {
        it('should render all children correctly', () => {
            expect(formRender).toMatchSnapshot();
        });

        it('should render null when children is not a function', () => {
            expect(shallow(
                <PassableProvider initialFormState={initialFormState}>
                    <input/>
                </PassableProvider>
            ).html()).toBe(null);
        });
    });

    describe('PassableProvider initialFormState prop', () => {
        it('should merge `field_1` initialState value into provider state', () => {
            expect(formRender.find('.username input').props().defaultValue).toBe('hello');
        });

        it('should merge `field_2` initialState value into provider state', () => {
            expect(formRender.find('.is-alive input').props().defaultChecked).toBe(true);
        });
    });
});
