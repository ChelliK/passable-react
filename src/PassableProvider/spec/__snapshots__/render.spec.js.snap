// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PassableProvider render function PassableProvider children rendering should render all children correctly 1`] = `
<form
  action=""
>
  <label
    className="username"
  >
    Username
    <input
      defaultValue="hello"
      name="field_1"
      type="text"
    />
  </label>
  <label
    className="is-alive"
  >
    Is Alive
    <input
      defaultChecked={true}
      name="field_2"
      type="checkbox"
    />
  </label>
</form>
`;
